File: algorithms/bst.rb
class Node
  attr_accessor :val, :left, :right, :parent

  def initialize(val, hash = {})
    @val    = val
    @left   = hash[:left]   if hash[:left]
    @right  = hash[:right]  if hash[:right]
    @parent = hash[:parent] if hash[:parent]
  end
end

class BST
  attr_accessor :root

  def initialize(val)
    @root = Node.new(val)
  end

  def add_left(node, val)
    node.left = Node.new(val, parent: node)
  end

  def add_right(node, val)
    node.right = Node.new(val, parent: node)
  end

  def show_ordered(node)
    if node != nil
      show_ordered(node.left)
      puts(node.val)
      show_ordered(node.right)
    end
  end

  def show(node)
    puts node.val
    show(node.left) if node.left
    show(node.right) if node.right
  end
end

bst = BST.new(20)

root = bst.root
left = bst.add_left(root, 15)
right = bst.add_right(root, 25)

bst.add_left(left, 13)
bst.add_right(left, 17)

bst.add_left(right, 23)
bst.add_right(right, 27)

bst.show_ordered(bst.root)

File: py/bst.py
import math

class Node():
  def __init__(self, val, **args):
    self.val = val
    if 'left' in args.keys():
      self.left = args['left']
    if 'right' in args.keys():
      self.right = args['right']
    if 'parent' in args.keys():
      self.parent = args['parent']


class BST():
  def __init__(self, arr):
    self.root = self.__get_node(arr, 0, len(arr) - 1)

  def __get_node(self, arr, s, e):
    if s <= e and s >= 0 and e < len(arr):
      mid = (s + e) // 2
      left_node  = self.__get_node(arr, s, mid - 1)
      right_node = self.__get_node(arr, mid + 1, e)
      return Node(arr[mid], left = left_node, right = right_node)
    return None

  def show(self, node):
    print(node.val)
    if node.left != None:
      self.show(node.left)
    if node.right != None:
      self.show(node.right)

  def show2(self, node, depth=0):
    ret = ""

    # Print right branch
    if node.right != None:
        ret += self.show2(node.right, depth + 1)

    # Print own val
    ret += "\n" + ("    "*depth) + str(node.val)

    # Print left branch
    if node.left != None:
        ret += self.show2(node.left, depth + 1)

    return ret

bst = BST([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

print(bst.show2(bst.root))